### PR TITLE
Flow tenant through processors as a string

### PR DESCRIPTION
**Pull request by esnible**
This PR replaces https://github.com/jaegertracing/jaeger/pull/3659 which passed the tenant ID string as part of a context.  It is intended as the first part of a true replacement for the proof-of-concept https://github.com/jaegertracing/jaeger/pull/3605

cc @pavolloffay

### CODE REVIEW

**Comment by pavolloffay** [comment]
@esnible is the intention here to make the writers tenant aware or have a writer per tenant and route to the appropriate writer in the span processor?

**Comment by esnible** [comment]
@pavolloffay The intention is to have the writers be tenant-aware.

@yurishkuro said "I see benefits to having per-tenant queues for better isolation" but later said "My suggestion would be to start with the simplest form: everything identical and shared, but tenancy exists at the data level."  This is an attempt to get the tenant down to the SpanWriter with as little churn as possible as a first PR.

File: storage/spanstore/interface.go [hunk 1]
```diff
@@ -86,3 +86,11 @@ type Operation struct {
 	Name     string
 	SpanKind string
 }
+
+// tenantKeyType is a custom type for the key "tenant", following context.Context convention
```
[the remark this answers is not in the document]
Comment (yurishkuro) [reply storage/spanstore/interface.go:90]: also, these should probably be defined at storage/ level, not storage/spanstore

File: storage/tenant.go [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+}
+
+// GetTenant retrieves a tenant associated with a Context
+func GetTenant(ctx context.Context) string {
+	tenant := ctx.Value(tenantKey)
+	if tenant == nil {
+		return ""
+	}
+
+	return tenant.(string)
```
Comment (yurishkuro) [inline storage/tenant.go:39]: perhaps we could make it safer
```go
if s, ok := tenant.(string) {
    return s
}
return ""
```

File: storage/tenant.go [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+// tenantKeyType is a custom type for the key "tenant", following context.Context convention
+type tenantKeyType string
+
+const (
+	// tenantKey holds tenancy for spans
+	tenantKey = tenantKeyType("tenant")
+)
+
+// WithTenant creates a Context with a tenant association
+func WithTenant(ctx context.Context, tenant string) context.Context {
```
Comment (yurishkuro) [inline storage/tenant.go:28]: let's add basic tests here to keep the coverage (and remove empty_test.go)

File: cmd/collector/app/span_processor_test.go [hunk 4]
```diff
@@ -566,3 +577,28 @@ func TestAdditionalProcessors(t *testing.T) {
...
+		Tenant: dummyTenant,
+	})
+	assert.NoError(t, err)
+	assert.Equal(t, []bool{true}, res)
+	assert.NoError(t, p.Close())
+
+	// Verify that the dummy tenant from SpansOptions context made it to writer
+	assert.Equal(t, true, w.tenants[dummyTenant])
+	// Verify no other tenantKey context values made it to writer
+	assert.True(t, reflect.DeepEqual(w.tenants, map[string]bool{dummyTenant: true}))
```
Comment (yurishkuro) [inline cmd/collector/app/span_processor_test.go:603]: I think you can simply use assert.Equal(), it should compare the maps correctly iirc.

**Review by yurishkuro — DISMISSED** [review]

**Review by yurishkuro — APPROVED** [review]

**Pull request merged**